python/dgl/nn/pytorch/explain/pgexplainer.py

```diff
@@ -0,0 +1,380 @@
+        else:
+            gate_inputs = torch.sigmoid(log_alpha)
+
+        return gate_inputs
+
+    def train_explanation_network(self, dataset, func_extract_feat):
+        r"""Training the explanation network by gradient descent(GD)
+         using Adam optimizer
+
+        Parameters
+        ----------
+        dataset : dgl.data
+            The dataset to train the importance edge mask.
+        func_extract_feat : func
+            A function that extracts the node embeddings for each individual graphs.
+        """
+        graph, _ = dataset[0]
+        self.model = self.model.to(graph.device)
+        self.elayers = self.elayers.to(graph.device)
+
+        optimizer = Adam(self.elayers.parameters(), lr=self.learning_rate)
+
+        ori_pred_dict = {}
+
+        self.model.eval()
+        with torch.no_grad():
+            for idx, (g, _) in enumerate(dataset):
+                feat = func_extract_feat(g)
+
+                logits = self.model(g, feat)
+                ori_pred_dict[idx] = logits.argmax(-1).data
+
+        # train the mask generator
+        for epoch in range(self.epochs):
+            loss = 0.0
+            pred_list = []
+
+            tmp = float(
+                self.init_tmp
+                * np.power(self.final_tmp / self.init_tmp, epoch / self.epochs)
+            )
+
+            self.elayers.train()
+            optimizer.zero_grad()
+
+            for idx, (g, _) in enumerate(dataset):
+                feat = func_extract_feat(g)
+
+                prob, edge_mask = self.explain_graph(
+                    g, feat, tmp=tmp, training=True
+                )
+
+                self.edge_mask = edge_mask
+
+                loss_tmp = self.loss(prob.unsqueeze(dim=0), ori_pred_dict[idx])
+                loss_tmp.backward()
+
+                loss += loss_tmp.item()
+                pred_label = prob.argmax(-1).item()
+                pred_list.append(pred_label)
+
+            optimizer.step()
+            print(f"Epoch: {epoch} | Loss: {loss}")
+
+    def explain_graph(self, graph, feat, tmp=1.0, training=False):
+        r"""Learn and return an an edge mask that play a crucial role to
+        explain the prediction made by the GNN for a graph. Also, return
+        the prediction made with just the edge mask.
+
+        Parameters
+        ----------
+        graph : DGLGraph
+            A homogeneous graph.
+        feat : Tensor
+            The input feature of shape :math:`(N, D)`. :math:`N` is the
+            number of nodes, and :math:`D` is the feature size.
+        tmp : float
+            The temperature parameter fed to the sample procedure
+        training : bool
+            Indicates whether the concrete_sample is called during
+            training or evaluation.
+
+        Returns
+        -------
+        Tensor, Tensor
+            The classification probability for graph with edge mask,
+            the probability mask for graph edges.
```

Comment:
What does "the probability mask" mean here?

Reply:
edge weights is a better term